pkg/pubsub/adapter/adapter.go

```diff
@@ -46,87 +66,66 @@ type Adapter struct {
-	sinkURI string
-
-	// transformerURI is the URI for the transformer.
-	// Used for channels.
-	transformerURI string
-
-	// extensions is the converted ExtensionsBased64 value.
-	extensions map[string]string
-
-	// converterType use to select which converter to use.
-	converterType converters.ConverterType
-
 	// reporter reports metrics to the configured backend.
 	reporter StatsReporter
 
 	// converter used to convert pubsub messages to CE.
 	converter converters.Converter
 
+	// projectID is the id of the GCP project.
+	projectID string
+
+	// namespacedName is the namespaced name of the resource this receive adapter belong to.
+	namespacedName types.NamespacedName
+
+	// resourceGroup is the resource group name this receive adapter belong to.
+	resourceGroup string
+
+	// args holds a set of arguments used to configure the Adapter.
+	args *AdapterArgs
+
 	// cancel is function to stop pulling messages.
 	cancel context.CancelFunc
 
 	logger *zap.Logger
-
-	// name of the resource this receive adapter belong to.
-	namespacedResourceName types.NamespacedName
 }
 
 // NewAdapter creates a new adapter.
 func NewAdapter(
 	ctx context.Context,
+	projectID clients.ProjectID,
+	namespace Namespace,
+	name Name,
+	resourceGroup ResourceGroup,
 	subscription *pubsub.Subscription,
 	outbound *nethttp.Client,
 	converter converters.Converter,
 	reporter StatsReporter,
-	namespace Namespace,
-	name Name,
-	sinkURI SinkURI,
-	transformerURI TransformerURI,
-	converterType converters.ConverterType,
-	extensions map[string]string) *Adapter {
+	args *AdapterArgs) *Adapter {
 	return &Adapter{
-		subscription:           subscription,
-		outbound:               outbound,
-		converter:              converter,
-		reporter:               reporter,
-		sinkURI:                string(sinkURI),
-		transformerURI:         string(transformerURI),
-		namespacedResourceName: types.NamespacedName{Name: string(name), Namespace: string(namespace)},
-		converterType:          converterType,
-		extensions:             extensions,
-		logger:                 logging.FromContext(ctx),
+		subscription:   subscription,
+		projectID:      string(projectID),
+		namespacedName: types.NamespacedName{Namespace: string(namespace), Name: string(name)},
+		resourceGroup:  string(resourceGroup),
+		outbound:       outbound,
+		converter:      converter,
+		reporter:       reporter,
+		args:           args,
+		logger:         logging.FromContext(ctx),
 	}
 }
 
-func (a *Adapter) Start(ctx context.Context) error {
+func (a *Adapter) Start(ctx context.Context, done func(error)) {
 	ctx, a.cancel = context.WithCancel(ctx)
-	defer a.cancel()
 
-	errCh := make(chan error, 1)
+	// Augment context so that we can use it to create CE attributes.
+	ctx = WithProjectKey(ctx, a.projectID)
+	ctx = WithTopicKey(ctx, a.args.TopicID)
+	ctx = WithSubscriptionKey(ctx, a.subscription.ID())
+
 	go func() {
-		errCh <- a.subscription.Receive(ctx, a.receive)
+		done(a.subscription.Receive(ctx, a.receive))
```

Comment:
I was following what was done in fanout https://github.com/google/knative-gcp/blob/master/pkg/broker/handler/fanout.go#L178
I had a more convoluted version before and Chen commented on..

You mean just do:
`return a.subscription.Receive(ctx, a.receive)` ?

I think it will be easier as is for writing the UTs... Unless you have a strong objection, I'd rather leave it for now... WDYT

Reply:
Yes, async callbacks are often an antipattern in go. A synchronous api can always be easily converted to an async callback by the caller using a goroutine.

Reply:
SG, done, thanks for the feeback